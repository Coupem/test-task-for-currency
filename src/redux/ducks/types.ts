import {
  LOAD_CURRENCY,
  LOAD_CURRENCY_FAILED,
  LOAD_CURRENCY_SUCCESS,
} from './actionTypes';

export interface ResponseCurrency {
  ccy: string;
  base_ccy: string;
  buy: string;
  sale: string;
}

export interface Currency {
  ccy: string;
  baseCcy: string;
  buy: string;
  sale: string;
}

export interface CurrencyState {
  currency: Currency[];
  isLoadingCurrency: boolean;
  error: string | null;
}

export interface FetchCurrencyFailedPayload {
  error: string;
}

export interface FetchCurrencyRequest {
  type: typeof LOAD_CURRENCY;
}

export type FetchCurrencySuccess = {
  type: typeof LOAD_CURRENCY_SUCCESS;
  payload: Currency[];
};

export type FetchCurrencyFailed = {
  type: typeof LOAD_CURRENCY_FAILED;
  payload: FetchCurrencyFailedPayload;
};

export type CurrencyActions =
  | FetchCurrencyRequest
  | FetchCurrencySuccess
  | FetchCurrencyFailed;

export interface StoreState {
  currency: CurrencyState;
}

// export interface PropSelector {
//   listCurrency: Currency[];
//   baseCurrency: string;
//   inputValue: string;
// }
