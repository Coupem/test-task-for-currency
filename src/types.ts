import { ChangeEventHandler } from 'react';

export const LOAD_CURRENCY_SUCCESS =
  'task-currency/currency/LOAD_CURRENCY_SUCCESS';

export const LOAD_CURRENCY = 'task-currency/currency/LOAD_CURRENCY';

export const LOAD_CURRENCY_FAILED =
  'task-currency/currency/LOAD_CURRENCY_FAILED';

export interface ResponseCurrency {
  ccy: string;
  // eslint-disable-next-line camelcase
  base_ccy: string;
  buy: string;
  sale: string;
}

export interface Currency {
  ccy: string;
  baseCcy: string;
  buy: string;
  sale: string;
}

export interface CurrencyState {
  currency: Currency[];
  isLoadingCurrency: boolean;
  error: string | null;
}

export interface RootState {
  currency: CurrencyState;
}

export interface FetchCurrencyFailedPayload {
  error: string;
}

export interface FetchCurrencyRequest {
  type: typeof LOAD_CURRENCY;
}

export type FetchCurrencySuccess = {
  type: typeof LOAD_CURRENCY_SUCCESS;
  payload: Currency[];
};

export type FetchCurrencyFailed = {
  type: typeof LOAD_CURRENCY_FAILED;
  payload: FetchCurrencyFailedPayload;
};

export type CurrencyActions =
  | FetchCurrencyRequest
  | FetchCurrencySuccess
  | FetchCurrencyFailed;

export interface StoreState {
  currency: CurrencyState;
}

export interface FavoriteState {
  ccy: string;
  buy: string;
  sale: string;
}

interface FavoriteStateForProp {
  ccy?: string;
  buy?: string;
  sale?: string;
}

export interface CurrencyStorage {
  [key: string]: FavoriteState;
}

export interface FavoriteCurrencyProp {
  [key: string]: FavoriteStateForProp;
}

export interface PropConverter {
  baseCurrency: string;
  convertedValue: number;
}

export interface PropsInput {
  baseCurrency: string;
  handleChangeCurrency: ChangeEventHandler<HTMLSelectElement>;
  handleChangeInput: (value: string) => void;
}

export type ErrorsHandler = {
  amount?: string;
};

export interface Primary {
  readonly primary: boolean;
}
