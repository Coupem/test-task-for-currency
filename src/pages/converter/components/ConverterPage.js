import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { ConverterContainer, TitleConverterContainer } from './styledComponets';
import InputGroupComponent from './InputGroupComponent';
import ConvertCurrencyComponent from './ConvertCurrencyComponent';

function ConverterPage({ loadedCurrency }) {
  const [valueInput, setInputValue] = useState('');
  const [convertedValueToNational, setConvertedValueToNational] = useState(1);
  const [baseCurrency, setBaseCurrency] = useState('UAH');

  const handleChangeBaseCurrency = (currency) => {
    setBaseCurrency(currency);
  };

  const handleInputValue = (event) => {
    const inputValue = Math.max(0, parseInt(event.target.value, 10))
      .toString()
      .slice(0, 10);
    setInputValue(inputValue);
  };

  const filteredCurrency = loadedCurrency.filter(
    (itemCurrency) => itemCurrency.ccy !== baseCurrency
  );

  useEffect(() => {
    let countValue = 1;

    if (baseCurrency === 'BTC') {
      // converted to USD
      const { buy } = loadedCurrency.find(
        (itemCurrency) => itemCurrency.ccy === 'USD'
      );
      countValue = valueInput * buy;
    }

    loadedCurrency.forEach((itemCurrency) => {
      if (baseCurrency === itemCurrency.ccy) {
        setConvertedValueToNational(countValue * valueInput * itemCurrency.buy);
      }
    });
  });

  return (
    <ConverterContainer>
      <TitleConverterContainer>Converter:</TitleConverterContainer>
      <InputGroupComponent
        loadedCurrency={loadedCurrency}
        baseCurrency={baseCurrency}
        handleChangeCurrency={handleChangeBaseCurrency}
        handleChangeInput={handleInputValue}
      />
      <ConvertCurrencyComponent
        listCurrency={filteredCurrency}
        convertedValue={convertedValueToNational}
      />
    </ConverterContainer>
  );
}

ConverterPage.propTypes = {
  loadedCurrency: PropTypes.arrayOf(
    PropTypes.shape({
      ccy: PropTypes.string,
      base_ccy: PropTypes.string,
      buy: PropTypes.string,
      sale: PropTypes.string,
    })
  ),
};

export default ConverterPage;
