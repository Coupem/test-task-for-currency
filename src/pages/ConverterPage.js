import React, { useState } from 'react';
import { useSelector } from 'react-redux';

import { ConverterContainer, TitleConverterContainer } from '../styled';
import { InputValuesGroup, ConvertCurrency } from '../components';
import { defaultCurrency } from '../constants';
import { getConvertToNational } from '../redux/ducks/selectors';

const ConverterPage = () => {
  const [inputValue, setInputValue] = useState('');
  const [baseCurrency, setBaseCurrency] = useState(defaultCurrency);

  const { isLoadingCurrency: isLoading } = useSelector(
    ({ responseData }) => responseData
  );
  const convertedValueToNational = useSelector(
    ({ responseData: { currency } }) =>
      getConvertToNational(currency, baseCurrency, inputValue)
  );

  const handleChangeBaseCurrency = (currency) => {
    setBaseCurrency(currency.target.value);
  };

  const handleInputValue = (value) => {
    setInputValue(value);
  };

  if (isLoading) {
    return null;
  }

  return (
    <ConverterContainer>
      <TitleConverterContainer>Converter:</TitleConverterContainer>
      <InputValuesGroup
        baseCurrency={baseCurrency}
        handleChangeCurrency={handleChangeBaseCurrency}
        handleChangeInput={handleInputValue}
        inputValue={inputValue}
      />
      <ConvertCurrency
        baseCurrency={baseCurrency}
        convertedValue={convertedValueToNational}
      />
    </ConverterContainer>
  );
};

ConverterPage.propTypes = {};

export default ConverterPage;
