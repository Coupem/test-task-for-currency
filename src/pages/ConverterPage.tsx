import React, { useState } from 'react';
import { useSelector } from 'react-redux';

import { string } from 'prop-types';
import { ConverterContainer, TitleConverterContainer } from '../styled';
import { InputValuesGroup, ConvertCurrency } from '../components';
import { defaultCurrency } from '../constants';
import { getConvertToNational } from '../redux/ducks/selectors';
import { Currency, CurrencyState } from '../redux/ducks/types';

const ConverterPage = (): JSX.Element => {
  const [inputValue, setInputValue] = useState('');
  const [baseCurrency, setBaseCurrency] = useState(defaultCurrency);

  const { isLoadingCurrency: isLoading } = useSelector(
    (currency: CurrencyState) => currency
  );
  const convertedValueToNational = useSelector(({ currency }: CurrencyState) =>
    getConvertToNational(currency, baseCurrency, inputValue)
  );

  const handleChangeBaseCurrency = (
    currency: React.ChangeEvent<HTMLSelectElement>
  ) => {
    setBaseCurrency(currency.target.value);
  };

  const handleInputValue = (value: string) => {
    setInputValue(value);
  };

  if (isLoading) {
    return null as any;
  }

  return (
    <ConverterContainer>
      <TitleConverterContainer>Converter:</TitleConverterContainer>
      <InputValuesGroup
        baseCurrency={baseCurrency}
        handleChangeCurrency={handleChangeBaseCurrency}
        handleChangeInput={handleInputValue}
        inputValue={inputValue}
      />
      <ConvertCurrency
        baseCurrency={baseCurrency}
        convertedValue={convertedValueToNational}
      />
    </ConverterContainer>
  );
};

ConverterPage.propTypes = {};

export default ConverterPage;
