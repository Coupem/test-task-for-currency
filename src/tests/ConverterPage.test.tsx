import { mount, ReactWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import configureStore from 'redux-mock-store';

import { Provider } from 'react-redux';
import ConverterPage from '../pages/ConverterPage';
import { RootState } from '../types';

describe('Converter page', () => {
  let state: RootState;
  let setUp;
  let component: ReactWrapper;

  beforeAll(() => {
    state = {
      currency: {
        currency: [
          { ccy: 'USD', baseCcy: 'UAH', buy: '27.05000', sale: '27.45000' },
          { ccy: 'EUR', baseCcy: 'UAH', buy: '31.85000', sale: '32.45000' },
          { ccy: 'UAH', baseCcy: 'UAH', buy: '1', sale: '1' },
          { ccy: 'RUR', baseCcy: 'UAH', buy: '0.35000', sale: '0.38000' },
          { ccy: 'BTC', baseCcy: 'USD', buy: '31125.3298', sale: '34401.6803' },
        ],
        isLoadingCurrency: false,
        error: 'string',
      },
    };
    const mockStore = configureStore();

    setUp = (store: RootState) =>
      mount(
        <Provider store={mockStore(store)}>
          <ConverterPage {...state} />
        </Provider>
      );

    component = setUp(state);
  });

  it('should render correctly', () => {
    expect(toJson(component)).toMatchSnapshot();
  });
});
