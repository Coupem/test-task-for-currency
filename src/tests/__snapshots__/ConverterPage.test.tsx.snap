// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Converter page should render correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <ConverterPage
    currency={
      Object {
        "currency": Array [
          Object {
            "baseCcy": "UAH",
            "buy": "27.05000",
            "ccy": "USD",
            "sale": "27.45000",
          },
          Object {
            "baseCcy": "UAH",
            "buy": "31.85000",
            "ccy": "EUR",
            "sale": "32.45000",
          },
          Object {
            "baseCcy": "UAH",
            "buy": "1",
            "ccy": "UAH",
            "sale": "1",
          },
          Object {
            "baseCcy": "UAH",
            "buy": "0.35000",
            "ccy": "RUR",
            "sale": "0.38000",
          },
          Object {
            "baseCcy": "USD",
            "buy": "31125.3298",
            "ccy": "BTC",
            "sale": "34401.6803",
          },
        ],
        "error": "string",
        "isLoadingCurrency": false,
      }
    }
  >
    <styled.div>
      <div
        className="sc-bqGGPW jzYdPK"
      >
        <styled.div>
          <div
            className="sc-dIvrsQ kIKOvq"
          >
            Converter:
          </div>
        </styled.div>
        <InputValuesGroup
          baseCurrency="UAH"
          handleChangeCurrency={[Function]}
          handleChangeInput={[Function]}
        >
          <Styled(Component)
            className="m-auto"
          >
            <ForwardRef(InputGroup)
              className="sc-fnVZcZ ceSpry m-auto"
            >
              <div
                className="sc-fnVZcZ ceSpry m-auto input-group"
              >
                <Formik
                  initialValues={
                    Object {
                      "amount": "",
                      "currency": "UAH",
                    }
                  }
                  onSubmit={[Function]}
                  validate={[Function]}
                >
                  <form>
                    <styled.input
                      data-testid="input"
                      name="amount"
                      onChange={[Function]}
                      placeholder="input value"
                      type="number"
                      value=""
                    >
                      <input
                        className="sc-hBMUJo gCIqZs"
                        data-testid="input"
                        name="amount"
                        onChange={[Function]}
                        placeholder="input value"
                        type="number"
                        value=""
                      />
                    </styled.input>
                    <styled.select
                      defaultValue="UAH"
                      name="currency"
                      onChange={[Function]}
                    >
                      <select
                        className="sc-bkbkJK bTnLfD"
                        defaultValue="UAH"
                        name="currency"
                        onChange={[Function]}
                      >
                        <styled.option
                          key="USD"
                          value="USD"
                        >
                          <option
                            className="sc-fFSPTT kWHaHm"
                            value="USD"
                          >
                            USD
                          </option>
                        </styled.option>
                        <styled.option
                          key="EUR"
                          value="EUR"
                        >
                          <option
                            className="sc-fFSPTT kWHaHm"
                            value="EUR"
                          >
                            EUR
                          </option>
                        </styled.option>
                        <styled.option
                          key="UAH"
                          value="UAH"
                        >
                          <option
                            className="sc-fFSPTT kWHaHm"
                            value="UAH"
                          >
                            UAH
                          </option>
                        </styled.option>
                        <styled.option
                          key="RUR"
                          value="RUR"
                        >
                          <option
                            className="sc-fFSPTT kWHaHm"
                            value="RUR"
                          >
                            RUR
                          </option>
                        </styled.option>
                        <styled.option
                          key="BTC"
                          value="BTC"
                        >
                          <option
                            className="sc-fFSPTT kWHaHm"
                            value="BTC"
                          >
                            BTC
                          </option>
                        </styled.option>
                      </select>
                    </styled.select>
                  </form>
                </Formik>
              </div>
            </ForwardRef(InputGroup)>
          </Styled(Component)>
        </InputValuesGroup>
        <ConvertCurrency
          baseCurrency="UAH"
          convertedValue={0}
        >
          <Styled(Component)
            bordered={true}
            hover={true}
            striped={true}
          >
            <ForwardRef
              bordered={true}
              className="sc-iemWCZ cunngp"
              hover={true}
              striped={true}
            >
              <table
                className="sc-iemWCZ cunngp table table-striped table-bordered table-hover"
              >
                <thead>
                  <tr>
                    <th>
                      Converted Value
                    </th>
                    <th>
                      Currency
                    </th>
                  </tr>
                </thead>
                <tbody>
                  <tr
                    key="USD"
                  >
                    <td
                      className="p-lg-3"
                      id="convertedValue"
                    >
                      0.000
                    </td>
                    <td
                      className="w-25 p-lg-2"
                      id="emblem"
                    >
                      USD
                    </td>
                  </tr>
                  <tr
                    key="EUR"
                  >
                    <td
                      className="p-lg-3"
                      id="convertedValue"
                    >
                      0.000
                    </td>
                    <td
                      className="w-25 p-lg-2"
                      id="emblem"
                    >
                      EUR
                    </td>
                  </tr>
                  <tr
                    key="RUR"
                  >
                    <td
                      className="p-lg-3"
                      id="convertedValue"
                    >
                      0.000
                    </td>
                    <td
                      className="w-25 p-lg-2"
                      id="emblem"
                    >
                      RUR
                    </td>
                  </tr>
                  <tr
                    key="BTC"
                  >
                    <td
                      className="p-lg-3"
                      id="convertedValue"
                    >
                      0.000
                    </td>
                    <td
                      className="w-25 p-lg-2"
                      id="emblem"
                    >
                      BTC
                    </td>
                  </tr>
                </tbody>
              </table>
            </ForwardRef>
          </Styled(Component)>
        </ConvertCurrency>
      </div>
    </styled.div>
  </ConverterPage>
</Provider>
`;
